### DISCUSSION TITLE
Make `AdditionalFiles` available during source generator Initialize step

### DISCUSSION THREAD

**Discussion by macon** [dotnet/roslyn#51266] [Ideas]
I'm trying to augment the protobuf/C# codegen. I want to pass a set of `AdditionalFiles` (pointing at my .protos) to my source generator and scan these to set up some SyntaxReceivers. Unfortunately I need to set up the SyntaxReceiver in the `Initialize` method of the source generator and at this point the AdditionalFiles are not available to me. The AdditionalFiles is only available in the `GeneratorExecutionContext` parameter of the `Execute` method.

It seems right to have all the config information that the consuming project is giving to the generator (e.g. AdditionalFiles) available at the Initialize step.

**Comment by ufcpp** [comment]
This is just WIP? `RegisterForAdditionalFileChanges` already exists but internal.

https://github.com/dotnet/roslyn/blob/230d0d0caa2ec3be6039202acdfdfecaeb083fd6/src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs#L120

**Comment by jasonmalinowski** [comment]
@macon Why do you want them in the initialize step? The model here is Initialize() you are just specifying how your generator works and it only runs once; all the actual work happens in Execute. But I can imagine you have some case where you could optimize your syntax walk if you had already processed the additional files first?

**Reply by ryzngard** [reply]
The natural follow up is: how does this change when files are added/removed between design time builds? Do we initialize all generators again?
